core/arch/arch.c

```diff
@@ -3355,45 +3355,59 @@ dr_mcontext_to_priv_mcontext(priv_mcontext_t *dst, dr_mcontext_t *src)
      * if we append to dr_mcontext_t in the future we'll need
      * to check src->size here.
      */
-    if (src->size != sizeof(dr_mcontext_t) &&
-        src->size != sizeof(dr_mcontext_t) - sizeof(dr_opmask_t))
+    if (src->size > sizeof(dr_mcontext_t))
         return false;
-    if (TESTALL(DR_MC_ALL, src->flags))
+    if (TEST(DR_MC_ALL, src->flags) &&
+        IF_X86_ELSE(src->size > offsetof(priv_mcontext_t, opmask), true)) {
         *dst = *(priv_mcontext_t *)(&MCXT_FIRST_REG_FIELD(src));
-    else {
-        if (TEST(DR_MC_INTEGER, src->flags)) {
+    } else {
+        if (TESTANY(DR_MC_INTEGER | DR_MC_ALL, src->flags)) {
```

Comment:
bug: This new code is true for *any* set of flags which is not correct behavior.  The original was correct: I'm not sure what this change is attempting to fix/improve?

Ditto below.

Reply:
I'm sorry. I should have checked the flag's value. I assumed DR_MC_ALL is a separate enum value that means "any" type. The DR_MC_ALL case was always covered by the quick full copy in the "if" part. Now it may still reach the "else" part if the client was old, even with DR_MC_ALL. In any case, I see now, thanks. Fixed.